homeassistant/components/konnected/config_flow.py

```diff
@@ -0,0 +1,732 @@
+        vol.Optional(CONF_POLL_INTERVAL, default=3): vol.All(
+            vol.Coerce(int), vol.Range(min=1)
+        ),
+    }
+)
+
+SWITCH_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_ZONE, "s_zone"): vol.In(ZONES),
+        vol.Optional(CONF_NAME): cv.string,
+        vol.Optional(CONF_ACTIVATION, default=STATE_HIGH): vol.All(
+            vol.Lower, vol.Any(STATE_HIGH, STATE_LOW)
+        ),
+        vol.Optional(CONF_MOMENTARY): vol.All(vol.Coerce(int), vol.Range(min=10)),
+        vol.Optional(CONF_PAUSE): vol.All(vol.Coerce(int), vol.Range(min=10)),
+        vol.Optional(CONF_REPEAT): vol.All(vol.Coerce(int), vol.Range(min=-1)),
+    }
+)
+
+DEVICE_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_ID): cv.matches_regex("[0-9a-f]{12}"),
+        vol.Optional(CONF_BINARY_SENSORS): vol.All(
+            cv.ensure_list, [BINARY_SENSOR_SCHEMA]
+        ),
+        vol.Optional(CONF_SENSORS): vol.All(cv.ensure_list, [SENSOR_SCHEMA]),
+        vol.Optional(CONF_SWITCHES): vol.All(cv.ensure_list, [SWITCH_SCHEMA]),
+        vol.Required(CONF_HOST): cv.string,
+        vol.Required(CONF_PORT): cv.port,
+        vol.Optional(CONF_BLINK, default=True): cv.boolean,
+        vol.Optional(CONF_DISCOVERY, default=True): cv.boolean,
+    }
+)
+
+
+@callback
+def configured_hosts(hass):
+    """Return a set of the configured hosts."""
+    return set(
+        entry.data.get("host") for entry in hass.config_entries.async_entries(DOMAIN)
+    )
+
+
+@callback
+def configured_devices(hass):
+    """Return a set of the configured devices."""
+    return set(entry.data["id"] for entry in hass.config_entries.async_entries(DOMAIN))
+
+
+class KonnectedFlowHandler(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for NEW_NAME."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_PUSH
+
+    # pylint: disable=no-member # https://github.com/PyCQA/pylint/issues/3167
+
+    def __init__(self):
+        """Initialize the Konnected flow."""
+        self.host = None
+        self.port = None
+        self.model = KONN_MODEL
+        self.device_id = None
+
+        # data above builds the config entry
+        # data below maintains state between steps
+        self.io_cfg = {}
+        self.binary_sensors = []
+        self.sensors = []
+        self.switches = []
+        self.active_cfg = None
+
+    @property
+    def cached_config(self):
+        """Retrieve cached config for device."""
+        if self.hass.data.get(DOMAIN) and self.hass.data[DOMAIN].get("config_data"):
+            return self.hass.data[DOMAIN]["config_data"].get(self.device_id)
+        return None
+
+    @callback
+    def async_cache_config(self, config):
+        """Cache the device config for shared usage."""
+        if DOMAIN not in self.hass.data:
+            self.hass.data[DOMAIN] = {}
+            self.hass.data[DOMAIN]["config_data"] = {}
+
+        self.hass.data[DOMAIN]["config_data"][self.device_id] = config
```

Comment:
Right now the pure config data goes into the cfg entry. hass.data has a combined config/state object which is generated when we synchronize settings with the device.  This mechanism was held over from the original implementation to avoid changing more code and introducing breaking changes.  It seems the hass state machine and config entry info should be able to replace this custom mechanism.  I will look to replace.